src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Tree/ProjectImports/ProjectImportsSubTreeProvider.cs

```diff
@@ -0,0 +1,330 @@
+                                    TearDownTree();
+                                }
+                            }
+                        });
+
+                    return;
+
+                    void SetUpTree()
+                    {
+                        Assumes.Null(_subscriptions);
+
+                        // Set a visible root
+                        SubmitTreeUpdateAsync(
+                            (currentTree, configuredProjectExports, token) =>
+                            {
+                                // Update (make visible) or create a new tree if no prior one exists
+                                IProjectTree tree = currentTree == null
+                                    ? NewTree(Resources.ImportsTreeNodeName, icon: s_rootIcon, flags: s_projectImportsTreeRootFlags)
+                                    : currentTree.Value.Tree.SetVisible(true);
+
+                                return Task.FromResult(new TreeUpdateResult(tree));
+                            });
+
+                        // Subscribe to data to populate the tree and keep it updated with changes
+                        IReceivableSourceBlock<IProjectVersionedValue<IProjectImportTreeSnapshot>> importTreeSource
+                            = _projectSubscriptionService.ImportTreeSource.SourceBlock;
+
+                        IReceivableSourceBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>> projectRuleSource
+                            = _projectSubscriptionService.ProjectRuleSource.SourceBlock;
+
+                        var intermediateBlock =
+                            new BufferBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>>(
+                                new ExecutionDataflowBlockOptions
+                                {
+                                    NameFormat = "Import Tree Intermediate: {1}"
+                                });
+
+                        _subscriptions ??= new DisposableBag();
+
+                        _subscriptions.Add(
+                            projectRuleSource.LinkTo(
+                                intermediateBlock,
+                                ruleNames: ConfigurationGeneral.SchemaName,
+                                suppressVersionOnlyUpdates: false,
+                                linkOptions: DataflowOption.PropagateCompletion));
+
+                        ITargetBlock<IProjectVersionedValue<ValueTuple<IProjectImportTreeSnapshot, IProjectSubscriptionUpdate>>> actionBlock =
+                            DataflowBlockSlim.CreateActionBlock<IProjectVersionedValue<ValueTuple<IProjectImportTreeSnapshot, IProjectSubscriptionUpdate>>>(
+                                SyncTree,
+                                new ExecutionDataflowBlockOptions
+                                {
+                                    NameFormat = "Import Tree Action: {1}"
+                                });
+
+                        using (TrySuppressExecutionContextFlow())
```

Comment:
I don't know what execution context flow is, or why it should be suppressed. Is there doco for this you know of to point me to?

Reply:
This pattern originated in CPS's capabilities tree provider.

[Above](https://github.com/dotnet/project-system/pull/5639#discussion_r345527552), @davkean asked @lifengl whether this is still needed.